### PR TITLE
Change flattening code to be iterative and not recursive

### PR DESCRIPTION
**Pull request by yanscalyr**
Rewrite of the hash/array flattening util code to be iterative, inspired by running into issues related to very deeply nested objects.

I suspect there are multiple ways to make this code nicer, but I was having difficulty keeping these stacks in my head as is (apparently my brain has less memory than what we give this plugin by quite a lot).

### CODE REVIEW

File: lib/scalyr/common/util.rb [hunk 1]
```diff
@@ -6,47 +6,80 @@ module Scalyr; module Common; module Util;
...
         end
-      else
-        result["#{index}"] = value
       end
-    end
+      result[result_key] = obj
 
-  else
+      if max_key_count > -1 and result.keys.count > max_key_count
+        raise RuntimeError.new("Resulting flattened object will contain more keys than the configured flattening_max_key_count of #{max_key_count}")
```
Comment (Kami) [inline lib/scalyr/common/util.rb:70]: I would if we could at least include some of the original event or similar to make it easier to correlate those errors to a specific event without enabling logstash debug log level which is very noisy.

**Review by Kami — APPROVED** [review]
Besides one comment, LGTM

File: lib/scalyr/common/util.rb (hunk 1 above)
Comment (yanscalyr) [reply lib/scalyr/common/util.rb:70]: A truncated sample of the event would do the trick I guess, same as in other parts of the code.

File: lib/scalyr/common/util.rb [hunk 2]
```diff
@@ -1,52 +1,96 @@
...
-    # input object is a hash
-    obj.each do |key, value|
-      if (flatten_arrays and value.respond_to?(:each)) or value.respond_to?(:has_key?)
-        flatten(value, fix_deep_flattening_delimiters ? delimiter : '_', flatten_arrays).each do |subkey, subvalue|
-          result["#{key}#{delimiter}#{subkey}"] = subvalue
-        end
-      else
-        result["#{key}"] = value
+  test_key = 0
+  #Debugging
```
Comment (Kami) [inline lib/scalyr/common/util.rb:36]: I guess you can remove those comments / debug statements :)

**Pull request merged**